proposals/0000-std-experimental.rst

```diff
@@ -0,0 +1,177 @@
+
+Semantically the last ``-[no-]experimental`` flag passed to GHC will win.
+
+Examples
+--------
+Consider a hypothetical experimental feature ``X``. Without the
+``-experimental`` flag, trying to use ``X`` would result in a compiler error.
+However, with the flag, the feature can be used, but with the understanding that
+it might change in future releases. ::
+
+  -- Without -experimental
+  ghc Main.hs  -- This will throw an error if Feature X is used
+
+  -- With -experimental
+  ghc -experimental Main.hs  -- This will compile successfully with Feature X
+
+  -- With -no-experimental
+  ghc -experimental -no-experimental Main.hs -- This will throw an error if Feature X is used
+
+  ghc -no-experimental Main.hs -package foo   -- This will throw an error if Feature X is used anywhere in Main.hs or the package foo.
+
+Effect and Interactions
+-----------------------
+The introduction of this flag provides a clear distinction between stable and
+experimental features. It ensures that developers are aware of the potential
+risks of using experimental features while providing a sandbox for testing out
+new GHC capabilities.
+
+Costs and Drawbacks
+-------------------
+- There might be an initial learning curve for developers to understand the
+  distinction between experimental and stable features.
+- Maintaining two sets of features (stable and experimental) might increase the
+  complexity of the compiler codebase.
+
+The authors believe the benefit of clearly separating experimental and those
+that adhere to a more rigourous change evolution easily offsets the costs for
+both end users and feature developers. For developers of experimental feature
+this also means they reserve the rigth to break them without warning in
+backwards incompatible ways at any moment in time leading to significantly
+lighter processes for experimental features.
+
+Prior Work
+----------
+The ``-experimental`` feature is similar to solutions for experimental
+features and implementation in other languages, and GHC would not have a
+bespoke solution here, but share a common approach with many other compilers.
+Do note that the provided approaches mix editions, channels, and feature gates.
+The goal is to illustrate that the concept of separating experimental features
+explicitly behind opt-in gates.
+
+- **Rust**: Rust has so called channels, which delineate stability by stable,
+  beta, and nightly. To opt-in locally to nightly features you'd use the
+  ``+nightly`` toolchain modifier. Effectively using a different Rust compiler
+  provided by ``rustup``.
+- **Java**: Java has the ``--enable-preview`` flag since Java 9, which allows
+  opt-in to experimental features.
+- **Gnu Compiler Collection**: GCC provides flags to opt-in to experimental
+  features through the ``-std=...`` flag. For example, it has flags like
+  ``-std=c++1z`` for experimental C++ features before they were standardized in
+  C++17.
+
```

Comment:
```suggestion
- **Kotlin**: Kotlin has a very fine grained mechanism for marking experimental features in APIs, so called *Opt-in requirements﻿*. Requirements can be either propagating (transitive) or non-propagating (non-transitive), where non-propagating features are used for internal parts and propagating features are meant for cases where these leak into dependent code. (Reference: [Opt-in requirements﻿](https://kotlinlang.org/docs/opt-in-requirements.html))
```

Reply:
I'm not sure if this needs to be added to this proposal. However, I think it might be worth to take a look at https://kotlinlang.org/docs/opt-in-requirements.html :smile_cat:

N.B. this relates to API features, not Compiler features.